src/main/java/pl/edu/mimuw/collections/Multiset.java

```diff
@@ -0,0 +1,130 @@
+package pl.edu.mimuw.collections;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+public class Multiset<E> implements IMultiset<E>, Iterable<E> {
+    
+    private Map<E, Integer> occurencesMap;
+    private int size;
+
+    public Multiset() {
+        this.size = 0;
+        this.occurencesMap = new HashMap<>();
+    }   
+    public int size() {
+        return size;
+    }
+    
+    public int count(E element) {
+        if(occurencesMap.containsKey(element)) {
+            return occurencesMap.get(element);
+        }
+        return 0;
+    }
+
+    public int add(E element, int occurrences) {
+        if(!occurencesMap.containsKey(element)) {
+            occurencesMap.put(element, occurrences);
+            return occurrences;
+        }
+        int prevOccurences = occurencesMap.get(element);
+        occurencesMap.put(element, occurrences + prevOccurences);
+        return occurrences + prevOccurences;
+    }
+
+    public boolean add(E element) {
+        if(!occurencesMap.containsKey(element)) {
+            occurencesMap.put(element, 1);
+            return true;
+        }
+        int prevOccurences = occurencesMap.get(element);
+        occurencesMap.put(element, prevOccurences + 1);
+        return true;
+    }
+
+    public boolean remove(E element) {
+        if(!occurencesMap.containsKey(element)) {
+            return false;
+        }
+        int prevOccurences = occurencesMap.get(element);
+        occurencesMap.put(element, Math.max(0, prevOccurences - 1));
+        return (Math.max(0, prevOccurences - 1) > 0);
+    }
+
+    public int remove(E element, int occurrences) {
+        if(!occurencesMap.containsKey(element)) {
+            return 0;
+        }
+        int prevOccurences = occurencesMap.get(element);
+        occurencesMap.put(element, Math.max(0, prevOccurences - occurrences));
+        return Math.max(0, prevOccurences - occurrences);
+    }
+
+    public Set<E> elementSet() {
+        Set<E> s = new HashSet<>();
+        for (Map.Entry<E, Integer> entry : occurencesMap.entrySet()) {
+            if(entry.getValue() != 0) s.add(entry.getKey());
+        }
+        return s;
+    }
+
+    public boolean contains(E element) {
+        if(!occurencesMap.containsKey(element)) {
+            if(occurencesMap.get(element) != 0) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    @Override
+    public boolean equals(Object object) {
+        if(object.getClass() == this.getClass()) {
+            return object.hashCode() == this.hashCode();
```

Comment:
o nie, tak tego się nie robi zdecydowanie, to herezja XD
gdybyś miał niezminnik, to wystarczyłoby
```
public boolean equals(Object o) {
  if (o == null) return false;
  if (o.getClass() != this.getClass()) return false;
  var m = (Multiset<?>) o;
  return Objects.equals(this.occurencesMap, o.occurencesMap);
}
```

Reply:
equals moze zwracać true tylko dla równych obiektów, a nie dla tych, ktore przypadkiem maja ten sam hash